types/client.d.ts

```diff
@@ -4,211 +4,171 @@ import { Duplex, Readable, Writable } from 'stream'
+    maxHeaderSize?: number;
+    /** The amount of time the parser will wait to receive the complete HTTP headers (Node 14 and above only). Default: `30e3` milliseconds (30s). */
+    headersTimeout?: number;
+  }
+
+  export interface DispatchOptions {
+    path: string;
+    method: string;
+    /** Default: `null` */
+    body?: string | Buffer | Uint8Array | Readable | null;
+    /** Default: `null` */
+    headers?: IncomingHttpHeaders | null;
+    /** The timeout after which a request will time out, in milliseconds. Monitors time between receiving a complete headers. Use 0 to disable it entirely. Default: `30e3` milliseconds (30s). */
+    headersTimeout?: number;
+    /** The timeout after which a request will time out, in milliseconds. Monitors time between receiving a body data. Use 0 to disable it entirely. Default: `30e3` milliseconds (30s). */
+    bodyTimeout?: number;
+    /** Whether the requests can be safely retried or not. If `false` the request won't be sent until all preceeding requests in the pipeline has completed. Default: `true` if `method` is `HEAD` or `GET`. */
+    idempotent?: boolean;
+  }
+
+  export interface RequestOptions extends DispatchOptions {
+    opaque?: unknown;
+    /** Default: `null` */
+    signal?: AbortSignal | EventEmitter | null;
+  }
+
+  export interface PipelineOptions extends RequestOptions {
+    /** `true` if the `handler` will return an object stream. Default: `false` */
+    objectMode?: boolean;
+  }
+
+  export interface UpgradeOptions extends DispatchOptions {
+    /** A string of comma separated protocols, in descending preference order. Default: `'Websocket'` */
+    protocol?: string;
+  }
+
+  export interface ConnectOptions extends DispatchOptions {
+  }
+
+  export interface ResponseData {
+    statusCode: number;
+    headers: IncomingHttpHeaders;
+    body: Readable;
+    opaque?: unknown;
+  }
+
+  export interface StreamData {
+    opaque: unknown;
+    trailers: Record<string, unknown>;
+  }
+
+  export interface UpgradeData {
+    headers: IncomingHttpHeaders;
+    socket: Duplex;
+    opaque: unknown;
+  }
+
+  export interface ConnectData {
+    statusCode: number;
+    headers: IncomingHttpHeaders;
+    socket: Duplex;
+    opaque: unknown;
+  }
+
+  export interface StreamFactoryData {
+    statusCode: number;
+    headers: IncomingHttpHeaders;
+    opaque: unknown;
+  }
+  export type StreamFactory = (data: StreamFactoryData) => Writable
+
+  export interface PipelineHandlerData {
+    statusCode: number;
+    headers: IncomingHttpHeaders;
+    opaque: unknown;
+    body: Readable;
+  }
+
+  export type PipelineHandler = (data: PipelineHandlerData) => Readable
+
+  export interface DispatchHandlers {
+    /** Invoked before request is dispatched on socket. May be invoked multiple times when a request is retried when the request at the head of the pipeline fails. */
+    onConnect?(abort: () => void): void;
```

Comment:
```suggestion
    onConnect?(abort: AbortController['abort']): void;
```

Reply:
It’s not an abortcontroller

Reply:
Oh! This is a generic _abort_ not a AbortController.abort. Gotcha! Then can we still keep the comment but also remove the `type AbortController`?

Reply:
what comment?